### ISSUE TITLE
[FEATURE] Add support for "coordinate" data streaming for ColumnDataSource

### ISSUE THREAD

**Issue by charlesbluca** [bokeh/bokeh#11101]
**Is your feature request related to a problem? Please describe.**

Let's say I want to make a `MultiLine` plot using a `ColumnDataSource`, from which new data is streamed in:

```python
p = figure(plot_width=400, plot_height=400)

source = ColumnDataSource({
    "xs": [[1, 2, 3], [1, 2, 3]],
    "ys": [[4, 5, 6], [7, 8, 9]],
    "colors": ["firebrick", "navy"],
})

p.multi_line(xs="xs", ys="ys", line_color="colors", source=source)
```

I have 2 lines and would like to stream additional coordinates to them. However, `ColumnDataSource.stream()` only allows data to be appended directly to a column's values. This means that new coordinates cannot be added to preexisting lines; only new lines can be added to the overall plot:

```python
>>> source.stream({"xs": [[4], [4]], "ys": [[7], [10]]})
ValueError: Must stream updates to all existing columns (missing: colors)

>>> source.stream({"xs": [[1, 2, 3]], "ys": [[10, 11, 12]], "colors": ["firebrick"]})

>>> source.data
{'xs': [[1, 2, 3], [1, 2, 3], [1, 2, 3]],
 'ys': [[4, 5, 6], [7, 8, 9], [10, 11, 12]],
 'colors': ['firebrick', 'navy', 'firebrick']}
```

Without this functionality, I am unable to stream new coordinates to my original `MultiLine` plot through a `ColumnDataSource`.

**Describe the solution you'd like**

A separate streaming function to stream new coordinates to preexisting lines, effectively making the above attempt work:

```python
>>> source.stream({"xs": [[4], [4]], "ys": [[7], [10]]})

>>> source.data
{'xs': [[1, 2, 3, 4], [1, 2, 3, 4]],
 'ys': [[4, 5, 6, 7], [7, 8, 9, 10]],
 'colors': ['firebrick', 'navy']}
```

**Describe alternatives you've considered**

I've tried to manually append the coordinates to the desired lines in the `ColumnDataSource`, but that doesn't result in updates on the plot.

Another way to achieve a multi line plot with data streaming is to use several `Line` plots on the same figure, all using their own columns in the `ColumnDataSource`:

```python
p = figure(plot_width=400, plot_height=400)

source = ColumnDataSource({
    "x": [1, 2, 3],
    "y1": [4, 5, 6],
    "y2": [7, 8, 9]
})

p.line(x="x", y="y1", line_color="firebrick", source=source)
p.line(x="x", y="y2", line_color="navy", source=source)
```

This is a pretty good alternative for smaller, simpler plots. However, this can get slow in use cases with more (30+) lines, and prevents the use of columns for metadata to be used in a `HoverTool`.

**Additional context**

I ran into this problem trying to make multi-line time series in the Dask Distributed dashboard (dask/distributed#4627); I ended up going with the alternative strategy stated above in that case, and this feature would enable the use of `HoverTool` in that scenario.

**Comment by bryevdv** [comment]
This  has definitely been asked for in the past. Given that it actually would not ever make sense to:

* stream to the end of the "top-level" arrays of a CDS, and
* stream to the ends of sub-arrays

at the same time [1], I think I am inclined to propose a new `multi_stream` API for this use-case. Making new API would also isolate the change and reduce risk. 

This would also be immediately useful for streaming to "patches" though that seems like a less common use case. I don't think this work should attempt to include streaming to images. 

cc @bokeh/dev for thoughts

[1] It's not possible at the same time because streaming to the "normal" columns would increase their length, while mulit-streaming to sub-arrays would *not* increase the length of the top-level array, violating the same-length assumption for CDS columns.

**Comment by bryevdv** [comment]
Just to make the proprosal more concrete:

Given
```
source.data = {
    'xs':  [[1,2,3], [4,5,6,7]]
    'ys':  [[9,8,7], [6,5,4,3]]
    'foo': [10, 20]
} 
```

I would propose:
```
source.multi_stream({
    'xs': [4, 9], 
    'ys': [10, 12],
})
```

The main risk for users is that presumably it would usually only make sense to multi_stream to all "multi" columns at once, but we don't have any good ways to check/enforce or warn if not.

Also note that the above requires streaming to *all* of the sub-sequences at once. i.e. you would have to stream to all the lines in a multi-line, you could not stream only to one of the sub-lines. This is certainly a limitation, but I think getting around it would require introducing some pretty ugly syntax (one the level of patch "slices"). Open to suggestions, though.

**Comment by mrocklin** [comment]
Checking in @charlesbluca do you need multi-line, or would two lines work for you?  One figure can contain many different plotting elements.

**Comment by charlesbluca** [comment]
I think `MultiLine` is the best option for Dask's case if we anticipate a worker/line count over maybe 15-20 (the example I listed is just a simplified version of the problem). From my understanding, Bokeh degrades in performance with large glyph counts, and each individual `line` plot on a figure counts as a glyph, whereas a `MultiLine` plot of several lines counts as just one.

Another more superficial reason for wanting a `MultiLine` plot is that it allows us to format the `ColumnDataSource` in a way which makes incorporating hover tools trivial:

```python
source = ColumnDataSource({
    "cpu": [[...], [...]],
    "time": [[...], [...]],
    "cpu_max": [100, 35],  # we can refer to this in a tool with @cpu_max

})
```

dask/distributed#4627 does use the strategy you're describing, and it seems to work well enough for the 8 workers I tested it with.

**Comment by bryevdv** [comment]
> From my understanding, Bokeh degrades in performance with large glyph counts,

This is true, but for more concrete context, in my experience, 20 glyphs is fine. 200 is not. :) Regardless I think this is a useful and sensible addition to make.

**Comment by mrocklin** [comment]
Ah, if we're thinking about a line per worker I'll push back.  I think that
we should try for visuals that scale up to larger numbers of workers.  I
wonder if, instead, we might choose something like plotting quantiles (min,
-1 sigma, mean, +1 sigma, max).  I think that that will provide a smoother
experience long term, and be robust to large numbers of workers.  Bokeh
probably has some nice options to visualize distributions like this.

**Comment by charlesbluca** [comment]
That makes sense to me - would you mind roughly paraphrasing that in dask/distributed#4668 so we can continue the discussion there?

**Comment by mrocklin** [comment]
Done

**Comment by charlesbluca** [comment]
Revisiting this issue after some time, I'm interested in what the handling for pandas Series or DataFrames should look like; if we make a multi-line CDS using a DataFrame, we see that there's an `index` column corresponding to the index of the DataFrame:

```python
>>> data = pd.DataFrame({"a" : [[1],[2]], "b": [[3],[4]]})
>>> cds = ColumnDataSource(data)
>>> cds.data
{'index': array([0, 1]),
 'a': array([list([1]), list([2])], dtype=object),
 'b': array([list([3]), list([4])], dtype=object)}
```

If we wanted to do a multi-stream to this CDS using a Series or DataFrame to supply the data

```python
>>> new_data = pd.Series({"a": [[2], [3]], "b": [[4], [5]]})
```

they would also have an `index` column that would inherently be part of the new data; what should be done with it? As I see:

- We can't stream to the CDS's index column in the conventional way, as that would result in uneven columns:

```python
>>> cds.multi_stream(new_data)  # this shouldn't be possible
>>> cds.data
{'index': array([0, 1, 0, 1]),
 'a': array([list([1, 2]), list([2, 3])], dtype=object),
 'b': array([list([3, 4]), list([4, 5])], dtype=object)}
```

- We could silently turn the data in the `index` column into a nested list, though that might cause some confusion:

```python
>>> cds.multi_stream(new_data)
>>> cds.data
{'index': array([[0, 1], [0, 1]]),
 'a': array([list([1, 2]), list([2, 3])], dtype=object),
 'b': array([list([3, 4]), list([4, 5])], dtype=object)}
```

- We could ignore the `index` column in multi-streams if we know that the new data is a Series or DataFrame:

```python
>>> cds.multi_stream(new_data)
>>> cds.data
{'index': array([0, 1]),
 'a': array([list([1, 2]), list([2, 3])], dtype=object),
 'b': array([list([3, 4]), list([4, 5])], dtype=object)}
```

- Or we could only allow dictionaries as input for new data here, completely preventing the situation.

I'm partial to the third option, though I'm interested in other's thoughts, especially on if this is a completely avoidable problem.

**Comment by bryevdv** [comment]
I think it has to be option 3 (or 4, but really we should support data frames in some way). Recall the distinction for a new `multi_stream`is that it only add to *sub*-arrays, and does not ever change the "top-level" column lengths. So that leaves two options I can think of:

* automatically detect which columns have sub-arrays, and only attempt to stream into those. We would just need to be very clear and explicit in the documentation on this point. 
* Make users be explicit which columns they want to stream into:
  ```
  source.multi_stream(new_data, for=("a", "b"))
  ```
  And then raise an error of the asked-for columns are not already suitable
